File: src/main/scala/com/backfield/imagegenerator/scala/sep2014/ImageGenerator.scala
package com.backfield.imagegenerator.scala.sep2014

import java.awt.image.BufferedImage
import java.io.File
import javax.imageio.ImageIO

object ImageGenerator {

  val squareSize = 20

  val numberOfSquares = 10

  val draw : Map[Symbol, (BufferedImage, Int, Int, Int) => Unit] = Map(
    'SQUARE -> ((image, x, y, color) => {
      for (squareX <- 0 until squareSize; squareY <- 0 until squareSize) {
        image.setRGB(x * squareSize + squareX, y * squareSize + squareY, color)
      }
    }),
    'UPPERLEFTTRIANGLE -> ((image, x, y, color) => {
      for (squareX <- 0 until squareSize; squareY <- 0 until (squareSize - squareX)) {
        image.setRGB(x * squareSize + squareX, y * squareSize + squareY, color)
      }
    }),
    'LOWERLEFTTRIANGLE -> ((image, x, y, color) => {
      for (squareX <- 0 until squareSize; squareY <- squareX until squareSize) {
        image.setRGB(x * squareSize + squareX, y * squareSize + squareY, color)
      }
    }),
    'UPPERRIGHTTRIANGLE -> ((image, x, y, color) => {
      for (squareX <- 0 until squareSize; squareY <- 0 until squareX) {
        image.setRGB(x * squareSize + squareX, y * squareSize + squareY, color)
      }
    }),
    'LOWERRIGHTTRIANGLE -> ((image, x, y, color) => {
      for(squareX <- 0 until squareSize; squareY <- (squareSize - squareX) until squareSize) {
        image.setRGB(x * squareSize + squareX, y * squareSize + squareY, color)
      }
    })
  )

  val drawTypes = draw.keys.toList

  def main(args : Array[String]) : Unit = {
    val image = new BufferedImage(squareSize * numberOfSquares, squareSize * numberOfSquares, BufferedImage.TYPE_INT_RGB)
    for(x <- 0 until numberOfSquares; y <- 0 until numberOfSquares) {
      draw(
        drawTypes(
          Randomness().nextInt(
            drawTypes.size
          )
        )
      )(image, x, y, Color.nextColor)
    }
    val file = new File("test.png")
    ImageIO.write(image, "png", file)
  }

}

File: src/main/java/com/backfield/imagegenerator/java/sep2014/ImageGenerator.java
package com.backfield.imagegenerator.java.sep2014;

import com.backfield.imagegenerator.java.sep2014.drawables.*;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.util.*;

public class ImageGenerator {

    public static int squareSize = 20;

    public static int numberOfSquares = 10;

    private String name;

    private Map<String, Drawable> drawables = new HashMap<String, Drawable>();

    private List<String> drawableTypes = new LinkedList<String>();

    public Random createRandom(String name) {
        Random random = new Random();
        long seed = 0;
        for (byte b : name.getBytes()) {
            seed += b;
        }
        random.setSeed(seed);
        return random;
    }

    public ImageGenerator(String name) {
        this.name = name;
        drawables.put(SquareDrawable.NAME, new SquareDrawable());
        drawableTypes.add(SquareDrawable.NAME);
        drawables.put(LowerLeftTriangleDrawable.NAME, new LowerLeftTriangleDrawable());
        drawableTypes.add(LowerLeftTriangleDrawable.NAME);
        drawables.put(LowerRightTriangleDrawable.NAME, new LowerRightTriangleDrawable());
        drawableTypes.add(LowerRightTriangleDrawable.NAME);
        drawables.put(UpperLeftTriangleDrawable.NAME, new UpperLeftTriangleDrawable());
        drawableTypes.add(UpperLeftTriangleDrawable.NAME);
        drawables.put(UpperRightTriangleDrawable.NAME, new UpperRightTriangleDrawable());
        drawableTypes.add(UpperRightTriangleDrawable.NAME);
    }

    public BufferedImage generateImage() {
        Random random = createRandom(this.name);
        Color color = new Color();
        BufferedImage image = new BufferedImage(squareSize * numberOfSquares, squareSize * numberOfSquares, BufferedImage.TYPE_INT_RGB);
        for(int x = 0; x < numberOfSquares; x++) {
            for(int y = 0; y < numberOfSquares; y++) {
                String key = drawableTypes.get(random.nextInt(drawableTypes.size()));
                drawables.get(key).draw(image, x, y, color.nextColor(random));
            }
        }
        image.flush();
        return image;
    }

    public static void main(String[] args) {
        ImageGenerator generator = new ImageGenerator("jbackfield@gmail.com");
        File file = new File("test.png");
        try {
            ImageIO.write(generator.generateImage(), "png", file);
        } catch(Exception e) {
            throw new RuntimeException(e);
        }
    }

}
